### ISSUE TITLE
Proposed RFC Feature: Project Manager Export Buttons

### ISSUE THREAD

**Issue by tkothadev** [o3de/sig-ui-ux#115]
### Summary:
Although O3DE users have a straight-forward, accessible and customizable way to export their projects via the export-project O3DE CLI command, no equivalent exists in the main user interface of O3DE. This RFC proposes changes to the existing Project Buttons in the Project Manager’s main page, whereby users can access the power and capabilities of the export system.

### What is the relevance of this feature?
Currently the only way to export projects is through the O3DE CLI tools or by manually handling export. This is inaccessible for new or non-technical users, and should ideally be grouped in a sensible place where all projects are managed.

This allows users to take full advantage of O3DE’s export-project functionality through an accessible and easy-to-use interface, without relying on a command line interface directly. This empowers more of the community to share O3DE Projects more readily.

### Feature design description:
The ethos for this feature is that the experience for exporting the project should be nearly identical to the existing process for building the project. The user flow should operate identically, with the only difference being that Project Manager runs the O3DE `export-project` command under the hood.

![image (4)](https://github.com/user-attachments/assets/2a948b87-3da0-48a7-9bdb-7e769f912108)
The user can access export using the dropdown from the project card with an additional option for editing the Export Configuration. These parameters are mapped directly to the existing parameters for the `export-project` command.

![image (3)](https://github.com/user-attachments/assets/915d42bc-be73-4cbe-b91b-40176c8cade6)
Upon choosing to export, the project card should show the current status of the export process by showing the last known log statement.

![image (2)](https://github.com/user-attachments/assets/1de61a5c-e488-41ba-8815-6db41efe88ca)
There is an additional option in the drop down menu for editing export configurations. The export configuration should pull up a menu very similar to the Android Project Generator, with the ability to edit the configuration details for any platform, along with the desired platform to export to (performed by selecting the targeted Export Script file). 

All necessary parameters can be collected from the O3DE settings specified in the o3de python codebase, and can be arranged in a manner like the figure on the left. Different sections group the parameters as logically as possible. The Save button writes the values in the form as new changes to the underlying configuration file for the O3DE project.

If there are platform specific parameters, they should be grouped by tabs at the bottom of the form. The exact styling of this configuration panel is considered a separate matter, as we follow the convention currently setup by the Android Project Generator. A followup design RFC is recommended for handling style guidelines for peripheral configuration panels in Python.

Taken altogether, the process for exporting should look something like the following:
![image](https://private-user-images.githubusercontent.com/112996779/353350141-629be47b-7288-4a26-bb9c-83ae05283f91.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzNjM2ODQsIm5iZiI6MTcyMjM2MzM4NCwicGF0aCI6Ii8xMTI5OTY3NzkvMzUzMzUwMTQxLTYyOWJlNDdiLTcyODgtNGEyNi1iYjljLTgzYWUwNTI4M2Y5MS5naWY_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDczMFQxODE2MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YTJhMjBmODNjMDFkZjI1NzgwOTU5MDY1MDdjMDY3MTkyNjljNThlMjcwMWUwMzkwNzMwMzE3Nzg5OWRkYzEwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.T2JE01RgW6sIGYG5ezTle048Ed9HHb8QGlB6-mdiPsY)

### Technical design description:
* In the Project Manager Source code, there already exists functionality to build projects, and render the Project Button as building is in progress, along with managing its task state in parallel with the Project Manager. This code can be duplicated and repurposed specifically for use with Exporting.
    * [ProjectButton](https://github.com/spham-amzn/o3de/blob/27551de8bda39cd9405b2ee328385c05ad5c52aa/Code/Tools/ProjectManager/Source/ProjectButtonWidget.cpp) will need new states for exporting, namely ProjectButtonState::Exporting and ProjectButtonState::ExportFailed
        * The Exporting state will use the same render logic as Building for in progress exporting, with only very minor cosmetic changes (mostly text)
    * ProjectExportController/Worker will facilitate the export process, in a very similar manner to [ProjectBuilderController](https://github.com/spham-amzn/o3de/blob/27551de8bda39cd9405b2ee328385c05ad5c52aa/Code/Tools/ProjectManager/Source/ProjectBuilderController.cpp)/[Worker](https://github.com/spham-amzn/o3de/blob/27551de8bda39cd9405b2ee328385c05ad5c52aa/Code/Tools/ProjectManager/Source/ProjectBuilderWorker.cpp)
        * [ProjectsScreen](https://github.com/spham-amzn/o3de/blob/27551de8bda39cd9405b2ee328385c05ad5c52aa/Code/Tools/ProjectManager/Source/ProjectsScreen.cpp) will own an instance of ProjectExportController, which in turn will instantiate ProjectExportWorker as needed
        * Because we use O3DE export-project API under the hood, the worker business logic can simply invoke the command line using QProcess, which allows us to execute the export-project as a CLI, and only pass in the project path. This works because the export-project command will populate unspecified parameters with the project's configuration details, which are saved to an external file, or with sensible defaults.

* For the configuration manager, the simplest method to implement is to add a panel similar to the Android Project Generator which allows us to directly query and edit the selected export script along with all of its necessary parameters. 
    * This can be implemented as a TkInter app that only needs to do the following:
        * Fetch the user’s current settings from the project .command_settings file 
        * Render all settings as shown in the configuration panel in the Feature description section.
        * The app includes a Save button to save all necessary changes.
        * Record the user’s desired Export Script to run. The Export Script will fetch the expected configuration parameters, or will find defaults if not specified. This has to be available for the export system to query. 
            * To implement this, the current export system should include the export script as a configuration parameter, which can be queried by export-project if the user does not specify one.

### What are the advantages of the feature?
* From a technical standpoint, because the export process is very similar to building, it simplifies the code changes needed. The same can be said for borrowing from Android Project Generator for the editing of export configurations.
* If users are used to the workflow for building projects, they will get acclimated to exporting quickly, because it is very similar
* It builds upon the existing infrastructure for export-project, giving it a front-end interface
* Maintenance is significantly easier. The Project Manager codebase is only responsible for sending a single parameter to the export script directly: the project path

### What are the disadvantages of the feature?
* Because it exposes certain export API related details via the configuration editor, users are expected to understand how the export system works. This is effective for developers quickly exporting from the O3DE UI, but a dedicated user testing session is recommended to gauge communication effectiveness
    * Tooltips with good descriptions can help alleviate this issue.

### How will this be implemented or integrated into the O3DE environment?
* The following code changes are expected:
    * Add a new Config Tool codebase similar to [Android Project Generator](https://github.com/spham-amzn/o3de/tree/27551de8bda39cd9405b2ee328385c05ad5c52aa/Code/Tools/Android/ProjectGenerator) but simplified, based on export configuration. The functionality to query and save parameters will be use the [command_utils](https://github.com/o3de/o3de/blob/development/scripts/o3de/o3de/command_utils.py).
    * Changes to Project Manager
        * Update the ProjectButtonWidget to include Export/Config functionality in the dropdown. Config option will invoke the Config Tool
        * Update PythonBindings to access all relevant Export and Logging API
        * Create ProjectExportController/Worker based on the pattern laid out by [ProjectBuilderController](https://github.com/spham-amzn/o3de/blob/27551de8bda39cd9405b2ee328385c05ad5c52aa/Code/Tools/ProjectManager/Source/ProjectBuilderController.cpp)/[Worker](https://github.com/spham-amzn/o3de/blob/27551de8bda39cd9405b2ee328385c05ad5c52aa/Code/Tools/ProjectManager/Source/ProjectBuilderWorker.cpp)
        * Update the ProjectScreen to include APIs like SetProjectExportSucces or SetProjectExportFail along with including the ProjectExportController

* Implementation will occur in two steps:
    * Step 1: Ensure the basic export option works correctly. For this the export script will just fetch the existing configuration.
        * There exists CLI functionality to edit the export configuration in the meantime
        * The export script configuration option will initially be hardcoded to the pc script
    * Step 2: Implement the configuration editor and associated option, and make sure the export option uses desired configuration.
* Both of these steps will show up in a single PR, and with one PR for one platform at a time. The ideal implementation route is that the solution works for all platforms, and followup PRs only address issues specific to each platform. 
    * Start with PC, then address platforms like Android seperately. iOS would need to be reviewed to see if it can follow this route, due to the current dependence on Xcode.

### Are there any alternatives to this feature?
* Utilize a dedicated UI workflow to guide users through each step of the export process. This may be useful for new users, but the task could get tiresome quickly in workflows requiring faster iteration cycles. 
* Manually build and export the project, but that gets tiresome for users quickly.
* Another possibility is to ignore the UI features altogether, and just use the existing CLI functionality. This can work with team setups that have automated builds and CI/CD environments, but having an existing route via UI would improve developer quality of life by not needing to resort to the terminal all the time.

### How will users learn this feature?
* There will be updated documentation in the project export section of the user manual for O3DE. 
* The feature is discoverable using the Project Button in the main page of Project Manager. The Export functionality will show up in the associated drop down menu, and fits nicely with existing options for handling projects. All necessary parameters for the export script will be automatically assigned by virtue of selecting the project, and fetching the right configuration details.

### Are there any open questions?
* Do we still desire a fully fleshed out export workflow/wizard from a UI or UX standpoint?
    * Currently the proposed feature set would act as a near perfect complement to the existing export functionality, but are there strong reasons to still consider a more extensive UI approach from a practical perspective? Such answers and proposals are out of scope for the current RFC, but are worth exploring for future initiatives.
    * One step in this direction could be to modify the export configuration to work in a manner similar to editing a Gem, through Project Manager’s Qt library.
* When the export task is finished, should Project Manager display the export directory in the file explorer?
